hfile_s3.c

```diff
@@ -1113,160 +1075,171 @@ static int write_authorisation_callback(void *auth, char *request, kstring_t *co
+    s3_auth_data *ad = (s3_auth_data *) adv;
 
-static int v4_auth_header_callback(void *ctx, char ***hdrs) {
-    s3_auth_data *ad = (s3_auth_data *) ctx;
-    char content_hash[HASH_LENGTH_SHA256];
-    kstring_t content = KS_INITIALIZE;
-    kstring_t authorisation = KS_INITIALIZE;
-    kstring_t token_hdr = KS_INITIALIZE;
-    char *date_html = NULL;
-    time_t now;
-    int idx;
+    ad->region.l = 0;
+    return kputsn(region->s, region->l, &ad->region) < 0;
+}
 
-    if (!hdrs) { // Closing connection
-        free_auth_data(ad);
-        return 0;
-    }
+//
+// Writing and reading handling
+//
 
-    now = time(NULL);
+// Some common code
 
-    if (update_time(ad, now)) {
-        return -1;
-    }
+#define S3_MOVED_PERMANENTLY 301
+#define S3_BAD_REQUEST 400
 
-    if (ad->creds_expiry_time > 0
-        && ad->creds_expiry_time - now < CREDENTIAL_LIFETIME) {
-        refresh_auth_data(ad);
-    }
 
-    if (!ad->id.l || !ad->secret.l) {
-        return copy_auth_headers(ad, hdrs);
-    }
+static struct {
+    kstring_t useragent;
+    CURLSH *share;
+    pthread_mutex_t share_lock;
+} curl = { { 0, 0, NULL }, NULL, PTHREAD_MUTEX_INITIALIZER };
+
+static void share_lock(CURL *handle, curl_lock_data data,
+                       curl_lock_access access, void *userptr) {
+    pthread_mutex_lock(&curl.share_lock);
+}
+
+static void share_unlock(CURL *handle, curl_lock_data data, void *userptr) {
+    pthread_mutex_unlock(&curl.share_lock);
+}
 
-    hash_string("", 0, content_hash, sizeof(content_hash)); // empty hash
 
-    ad->canonical_query_string.l = 0;
+static void initialise_authorisation_values(hFILE_s3 *fp) {
+    ks_initialize(&fp->content_hash);
+    ks_initialize(&fp->authorisation);
+    ks_initialize(&fp->content);
+    ks_initialize(&fp->date);
+    ks_initialize(&fp->token);
+    ks_initialize(&fp->range);
+}
 
-    if (ad->user_query_string.l > 0) {
-        kputs(ad->user_query_string.s, &ad->canonical_query_string);
 
-        if (order_query_string(&ad->canonical_query_string)) {
-            return -1;
+static void clear_authorisation_values(hFILE_s3 *fp) {
+    ks_clear(&fp->content_hash);
+    ks_clear(&fp->authorisation);
+    ks_clear(&fp->content);
+    ks_clear(&fp->date);
+    ks_clear(&fp->token);
+    ks_clear(&fp->range);
+}
+
+
+static void free_authorisation_values(hFILE_s3 *fp) {
+    ks_free(&fp->content_hash);
+    ks_free(&fp->authorisation);
+    ks_free(&fp->content);
+    ks_free(&fp->date);
+    ks_free(&fp->token);
+    ks_free(&fp->range);
+}
+
+/* As the response text is case insensitive we need a version of strstr that
+   is also case insensitive.  The response is small so no need to get too
+   complicated on the string search.
+*/
+static char *stristr(char *haystack, char *needle) {
+
+    while (*haystack) {
+        char *h = haystack;
+        char *n = needle;
+
+        while (toupper(*h) == toupper(*n)) {
+            h++, n++;
+            if (!h || !n) break;
```

Comment:
I think you wanted this here:
```suggestion
            if (!*h || !*n) break;
```